crates/bevy_pbr/src/material.rs

```diff
@@ -653,180 +696,173 @@ fn extract_mesh_materials<M: Material>(
+            _marker: Default::default(),
         }
+    }
+}
 
-        if depth_prepass {
-            view_key |= MeshPipelineKey::DEPTH_PREPASS;
-        }
+#[derive(Clone, Resource, Debug)]
+pub struct EntitySpecializationTicks<M> {
+    pub entities: MainEntityHashMap<Tick>,
+    _marker: PhantomData<M>,
+}
 
-        if motion_vector_prepass {
-            view_key |= MeshPipelineKey::MOTION_VECTOR_PREPASS;
+impl<M> Default for EntitySpecializationTicks<M> {
+    fn default() -> Self {
+        Self {
+            entities: MainEntityHashMap::default(),
+            _marker: Default::default(),
         }
+    }
+}
 
-        if deferred_prepass {
-            view_key |= MeshPipelineKey::DEFERRED_PREPASS;
-        }
+#[derive(Resource)]
+pub struct SpecializedMaterialPipelineCache<M> {
+    map: HashMap<(MainEntity, MainEntity), (Tick, CachedRenderPipelineId), EntityHash>,
+    marker: PhantomData<M>,
+}
 
-        if temporal_jitter {
-            view_key |= MeshPipelineKey::TEMPORAL_JITTER;
+impl<M> Default for SpecializedMaterialPipelineCache<M> {
+    fn default() -> Self {
+        Self {
+            map: HashMap::default(),
+            marker: PhantomData,
         }
+    }
+}
 
-        if has_environment_maps {
-            view_key |= MeshPipelineKey::ENVIRONMENT_MAP;
-        }
+impl<M> Deref for SpecializedMaterialPipelineCache<M> {
+    type Target = HashMap<(MainEntity, MainEntity), (Tick, CachedRenderPipelineId), EntityHash>;
 
-        if has_irradiance_volumes {
-            view_key |= MeshPipelineKey::IRRADIANCE_VOLUME;
-        }
+    fn deref(&self) -> &Self::Target {
+        &self.map
+    }
+}
 
-        if has_oit {
-            view_key |= MeshPipelineKey::OIT_ENABLED;
-        }
+impl<M> DerefMut for SpecializedMaterialPipelineCache<M> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.map
+    }
+}
 
-        if let Some(projection) = projection {
-            view_key |= match projection {
-                Projection::Perspective(_) => MeshPipelineKey::VIEW_PROJECTION_PERSPECTIVE,
-                Projection::Orthographic(_) => MeshPipelineKey::VIEW_PROJECTION_ORTHOGRAPHIC,
-                Projection::Custom(_) => MeshPipelineKey::VIEW_PROJECTION_NONSTANDARD,
-            };
-        }
+fn check_entities_needing_specialization<M>(
+    mut thread_queues: Local<Parallel<Vec<Entity>>>,
+    mut needs_specialization: Query<
+        Entity,
+        Or<(
+            Changed<Mesh3d>,
+            AssetChanged<Mesh3d>,
+            Changed<MeshMaterial3d<M>>,
+            AssetChanged<MeshMaterial3d<M>>,
+        )>,
+    >,
+    mut entities_needing_specialization: ResMut<EntitiesNeedingSpecialization<M>>,
+) where
+    M: Material,
+{
+    entities_needing_specialization.entities.clear();
+    needs_specialization.par_iter_mut().for_each_init(
+        || thread_queues.borrow_local_mut(),
+        |queue, entity| {
+            queue.push(entity.into());
+        },
+    );
```

Comment:
It's surprising to me that the parallelized version of this wins when there's so little work to do.

Reply:
Yeah, this was from an earlier version where there was more happening inside here. I'll remove since the coordination cost is almost certainly worse.